### DISCUSSION TITLE
Errors and wrinkles in HQDM as documented

### DISCUSSION THREAD

**Discussion by ClimbingAl** [gchq/MagmaCore#88] [General]
**Issues and recommended strategies for dealing with them for HQDM as originally documented**

The book _Developing High Quality Data Models_ contains, in Chapter 17, the model that the Magma Core repo implements.  The original model contains 231 entities and many relationships.  Type and relationship inheritance is vital for the enforcement of the model patterns that the book documents.  Some errors, omissions and wrinkles have been discovered and it's worth keeping track of them and their consequences.

**Comment by ClimbingAl** [comment]
**Relationship inheritance issue in entity types  `Aggregation`, `Composition` and `Temporal_Composition`**
`Aggregation` is a subtype of `relationship` (an entity type that is a reification of types of binary relations) that should inherit the `member_of` relationship that relates instances of `relationship` (and its subtypes) with `class_of_relationship` (or its corresponding subtype(s)).  This pattern is shown in section 17.3 of the book but the listed subtypes (documented in section 17.4) don't show the `member_of` relationship and a corresponding subtype of `class_of_relationship`.

**Comment by ClimbingAl** [comment]
**The entity type `class_of_representation` is missing a SUBTYPE OF statement**
In the book, section 17.45, `class_of_representation` is missing a statement "SUBTYPE OF (class_of_association)".  This is correct in Magma Core (possibly only in the *develop* branch) but, in case of using the book as a reference or using legacy code, this inheritance path is missing.

**Reply by jamrce** [reply]
To note, this has been rectified in the hqdmTop EXPRESS file by [PR #2](https://github.com/hqdmTop/hqdmFramework/pull/2).

**Comment by ClimbingAl** [comment]
The entity type [functional_system](https://github.com/hqdmTop/hqdmFramework/wiki/functional_system) inherits `intended_role` from *functional_object*.  However, the corresponding relation `intended_role_by_class` is missing from [kind_of_functional_system](https://github.com/hqdmTop/hqdmFramework/wiki/kind_of_functional_system) and from *kind_of_functional_system_component*.  From the linked diagram for *kind_of_functional_system* it can be seen that `intended_role_by_class` should be present as an equivalent commitment to `natural_role_by_class` for *kind_of_biological_system*.

The reason for this is seems to be that `intended_role_by_class` is introduced to [kind_of_functional_object](https://github.com/hqdmTop/hqdmFramework/wiki/kind_of_functional_object) but isn't propagated down to these *system* related kinds as they inherit from class_of_functional_X.  Addressing this formally in the model would require some careful analysis but using `intended_role_by_class` for *kind_of_functional_system* and *kind_of_functional_system_component* is consistent with the general *X_role_by_class* pattern documented by Matthew can be useful.

**Reply by jamrce** [reply]
Analysis has found the missing subtypes for the system and system component extensions are:
- kind_of_functional_system requires subtype: kind_of_ordinary_functional_object
- kind_of_functional_system_component requires subtype: kind_of_functional_object
- kind_of_biological_system requires subtype: kind_of_ordinary_biological_object
- kind_of_biological_system_component requires subtype: kind_of_biological_object
